packages/@aws-cdk/aws-stepfunctions-tasks/lib/sns/publish.ts

```diff
@@ -4,6 +4,49 @@ import * as sfn from '@aws-cdk/aws-stepfunctions';
 import { Construct } from 'constructs';
 import { integrationResourceArn, validatePatternSupported } from '../private/task-utils';
 
+/**
+ * SNS Message Attribute type enum
+ */
+export enum SnsMessageAttributeType {
+
+  /**
+   * String type attributes.
+   *
+   * @see https://docs.aws.amazon.com/sns/latest/dg/sns-message-attributes.html#SNSMessageAttributes.DataTypes
+   */
+  STRING = 'String',
+
+  /**
+   * Binary type attributes.
+   *
+   * @see https://docs.aws.amazon.com/sns/latest/dg/sns-message-attributes.html#SNSMessageAttributes.DataTypes
+   */
+  BINARY = 'Binary',
+}
+
+/**
+ * SNS Message Attribute
+ */
+export interface SnsMessageAttribute {
+  /**
+   * Name is the name of the message attribute
+   */
+  readonly name: string;
+  /**
+   * MessageAttributeType is the type of data being passed.
+   *
+   *
+   * @see https://docs.aws.amazon.com/sns/latest/dg/sns-message-attributes.html#SNSMessageAttributes.DataTypes
+   * @default STRING and StringValue will be set
```

Comment:
```suggestion
   * @default SnsMessageAttributeType.STRING
```

what does the StringValue refer to?

Reply:
StringValue is the type that must be set in the JSON States language. There are two primary types Binary/BinaryValue and String/StringValue. In addition to these in some languages the [API ](https://docs.aws.amazon.com/sns/latest/api/API_Publish.html) supports other [variations](https://docs.aws.amazon.com/sns/latest/api/API_MessageAttributeValue.html) I think based on use cases StepFunction is going to very rarely need to support binary. The most common need here based on the issue is filter policies based on strings. You can see this usage with `DataType` and `StringValue` in the examples [here](https://docs.aws.amazon.com/step-functions/latest/dg/connect-sns.html).

Reply:
it would be really cool if we could set this automatically based on the type of the value. if the value is from the input (JsonPath.isEncodedJsonPath) or unresolved (Token.isUnresolved) then we default to string, otherwise we check if the type is a string, number, or array. if none of the above (and arrays) we can serialize the value as well

Reply:
@BenChaimberg  - I have pushed a branch to demonstrate the issues with the request. I'll talk about the key points for `Binary` and `Token`:

`Binary`: IMHO I do not see a good use case for the support. I am interested if anybody can provide a good use case. I demonstrated that we could use [Typed Arrays](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray) to handle processing byte arrays and identifying them as Binary. I ran into a final issue converting those to strings and paused while this discussion happens (I think TextEncoder may be the answer). In order to use this function a developer would basically need to read in some binary data and shove it into the CloudFormation (it's constant at this point, plus or minus CFN condition magic). If I reached a situation where I needed to do that, I would put the object in S3 or Dynamo and use other means to access it. The more likely use case would be that some function dumps a chunk of binary data back into the state payload. This will either be a string array of size 1 or a base64 encoded string. I can't think of any way to introspect this is binary data during synth. For this reason I think if we must support `Binary` I would propose an optional attribute and moving back to a defined MessageAttribute Interface. Again, for the StepFunction use case IMO I think Binary is near obsolete.

`Token`: Will "work" in this current implementation due to all the amazing work that Rico's put into Token <-> String interops. However, I think that may prove problematic. The StepFunctions implementation as is *does not* support substitutions. I say that because of the work for [integrationResourceArn](https://github.com/aws/aws-cdk/blob/7966f8d48c4bff26beb22856d289f9d0c7e7081d/packages/%40aws-cdk/aws-stepfunctions-tasks/lib/private/task-utils.ts#L29-L35). Tokens are used pervasively in CDK, but some may expect a `Ref: <ResourceID>` to actually resolve and with this limitation it will silently fail until the user tests the SFN code via deploy. 

In summary I encourage CDK to not support Binary, but will disagree and commit by changing back to an interface. I am open to proposals on how best to handle `Token` and I did consider making it fail, but that felt too rigid.

Reply:
The issue is Token resolution, that part just works. The issue is the prior decision to not support [Step Function Substitution](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-stepfunctions-statemachine.html#cfn-stepfunctions-statemachine-definitionsubstitutions). 

Let's use an example with the current code, if I set up the following test case:

```ts
  test('with message attributes', () => {
    // GIVEN
    const stack = new cdk.Stack();
    const topic = new sns.Topic(stack, 'Topic');
    const stream = new Stream(stack, 'TokenExample');

    // WHEN
    const task = new SnsPublish(stack, 'Publish', {
      topic,
      message: sfn.TaskInput.fromText('Publish this message'),
      messageAttributes: {
        cake: 'chocolate',
        cakePic: new Uint8Array(16),
        cakeCount: 2,
        vendors: ['Great Cakes', 'Cakes R Us', 'Local Cakes'],
        stream: stream.streamArn,
      },
    });
```
The output render  the token for ARN correctly as:

```ts
 ... snip ...
     "stream": {
             "DataType": "String",
             "StringValue":  {
               "Fn::GetAtt":  [
                 "TokenExample182385A1",
                 "Arn",
               ],
             },
... snip ...
```

However because we don't define substitution correctly this either fails at deploy time or shoves that into a string and fails at runtime because the ARN doesn't actually get replaced correctly. The SFN definition substitution needs to get created and that needs to be `${SubstituteName}` instead of the CFN Fn::GetAtt function. If in fact the Token is just a string, then this just works. However, I feel like this gotcha will be a very bad user experience. A broader refactor that inspects for any `Token` in the StepFunction Definition and automatically sets the a substitution for it will work, but that's a much larger feature than this addition.